### PR TITLE
Add support for getting value as Period

### PR DESCRIPTION
**Pull request by kag0**
from issue #473 

I think `parseDuration` and `parsePeriod` could probably be re-worked to be a little more dry (something with `ChronoUnit.isTimeBased` rather than `TimeUnit.toNanos`), but didn't feel the need to do that now as this is a pretty superficial feature.

### CODE REVIEW

**Comment by kag0** [comment]
Signed cla

**Comment by havocp** [comment]
Thanks! Looking pretty good. We might consider a unit test in UnitParserTest.scala covering some of the weird corner cases (especially error handling), and maybe update HOCON.md to describe how periods are parsed differently from durations? there's a section in there for durations already.

**Comment by kag0** [comment]
@havocp how's that looking?

File: config/src/main/java/com/typesafe/config/impl/SimpleConfig.java [hunk 1]
```diff
@@ -322,6 +326,20 @@ public Duration getDuration(String path) {
         return Duration.ofNanos(nanos);
     }
 
+    @Override
+    public Period getPeriod(String path){
+        ConfigValue v = find(path, ConfigValueType.STRING);
+        return parsePeriod((String) v.unwrapped(), v.origin(), path);
+    }
+
+    public TemporalAmount getTemporal(String path){
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/SimpleConfig.java]: does this one need @Override?

File: config/src/main/java/com/typesafe/config/Config.java [hunk 2]
```diff
@@ -794,6 +796,44 @@
...
+     * @throws ConfigException.BadValue
+     *             if value cannot be parsed as a number of the given TimeUnit
+     */
+    Period getPeriod(String path);
+
+    /**
+     * Gets a value as a java.time.temporal.TemporalAmount.
+     * This method will first try get get the value as a java.time.Duration, and if unsuccessful,
+     * then as a java.time.Period.
+     * This means that values like "5m" will be parsed as 5 minutes rather than 5 months
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/Config.java]: I feel like the ambiguous values here ("m" and no units?) are fairly dangerous perhaps ... an option would be to outright prohibit them... though that could be confusing also. I'm OK with what you have here but do wonder if there's an option that doesn't leave a pitfall for whoever is writing the config.

Comment (kag0) [inline config/src/main/java/com/typesafe/config/Config.java]: I had that thought as well. I think it it can be confusing, but most applications will use either `getPeriod` or `getDuration` and will not need to worry. For me it's just balancing how many people will be confused by `m` being unavailable vs how many will use `getTemporal`.

File: config/src/main/java/com/typesafe/config/impl/SimpleConfig.java (hunk 1 above)
Comment (kag0) [inline config/src/main/java/com/typesafe/config/impl/SimpleConfig.java]: yes, let me add that

File: config/src/main/java/com/typesafe/config/Config.java (hunk 2 above)
Comment (gmethvin) [inline config/src/main/java/com/typesafe/config/Config.java]: `TemporalAmount` is not really meant to be used in application code. According to the [javadoc](https://docs.oracle.com/javase/8/docs/api/java/time/temporal/TemporalAmount.html):

> This interface is a framework-level interface that should not be widely used in application code. Instead, applications should create and pass around instances of concrete types, such as Period and Duration.

If you're going to force users to decide how to handle the `TemporalAmount` returned from this method, you might as well let them implement the relatively simple logic in `getTemporal` themselves, so they can decide how to handle the ambiguity of minutes/months.

In a practical sense, I think it would be more useful to have `getDuration` support weeks, months, and years as units using standard lengths for those units. In most cases the underlying API I'm using (e.g. max-age for a cookie) expects a precise number of nanoseconds, so a `Period` is usually not what I want anyway.

**Comment by kag0** [comment]
> I would also expect to be able to write
foo = P1Y2M3D
using the ISO-8601 period format.

Is there support currently for doing this with durations? If so then I think there is an argument for adding it. But if not, can we make it a separate issue? Presumably users would use the smallest applicable unit to describe the period they needed.

File: config/src/main/java/com/typesafe/config/Config.java (hunk 2 above)
Comment (kag0) [inline config/src/main/java/com/typesafe/config/Config.java]: I had assumed that the config library would be used by framework creators as well as directly by applications. Is there a convenient way we can surface the ability to get a `TemporalAmount` without putting it directly on the `Config` interface?

**Comment by gmethvin** [comment]
> Is there support currently for doing this with durations?

No, but that's not a problem because Durations are not based on the ISO-8601 calendar. `Duration` assumes a standard length for each of the time units involved. That's not the case for periods. The period "1 month and 15 days" could correspond to "45 days", "46 days", etc. depending on when on the calendar it starts. So for periods we need to store the individual fields.

That said, `Duration.parse` allows you to parse a format like "PT1H2M34S".

If the goal is simply to support a number with unit it would be far easier to use, and more semantically correct, if we returned a class representing a tuple of `Long` and `ChronoUnit`. `Period` is not the proper type to use.

**Comment by kag0** [comment]
> The period "1 month and 15 days" could correspond to "45 days", "46 days", etc.

A very good point. With that in mind, parsing ISO periods should definitely be supported. 
`P1Y2M3D`
`1 month`
`2 days`
`1y`
Should all be valid values.

File: config/src/main/java/com/typesafe/config/Config.java (hunk 2 above)
Comment (gmethvin) [inline config/src/main/java/com/typesafe/config/Config.java]: @kag0 can you describe a use case for a framework creator?

> Is there a convenient way we can surface the ability to get a TemporalAmount without putting it directly on the Config interface?

What does it mean to "get a TemporalAmount" to you?

Comment (kag0) [inline config/src/main/java/com/typesafe/config/Config.java]: In this case "get a TemporalAmount" to me just means getting either a `Duration` or `Period` but not knowing which. 

I could imagine a framework creator might have some setting where they might want to do `OffsetDateTime.now().plus(maintenanceReminderInterval)` and wouldn't want to constrain the user to a period or duration, but do want to prevent taking a `TemporalAmount` directly.

Comment (gmethvin) [inline config/src/main/java/com/typesafe/config/Config.java]: Yeah, I feel like in that case you're adding the possibility to return `Duration` as a hack because `Period` doesn't support units smaller than days. Ideally you want to be able to return a type that supports both, such as [PeriodDuration](http://www.threeten.org/threeten-extra/apidocs/org/threeten/extra/PeriodDuration.html) from threeten-extra.

**Comment by gmethvin** [comment]
Expanding on my earlier proposal, we could have a general method to parse a numeric value with a chronological unit. We could have a type like this:

```java
public final class ConfigChronoAmount {
  private final long amount; // maybe should be BigDecimal/double?
  private final ChronoUnit unit;
  // ... constructor, getters ...
}
```
and add a method `ConfigChronoAmount getChrono(String path)` that does similar parsing as `parseDuration` (perhaps consolidate the methods).

Note that it is trivial to convert this to a duration using the estimated values of the units (`unit.getDuration().multipliedBy(amount)`). I would argue that's the most common situation. If I set my timeout to `1 year` it's fine to use a value of 365 days instead of a different one depending on the calendar. The `year` is just a shorthand.

The only thing I'm not sure about is whether we should support fractional units like `1.5 months`. It would be nice to do that from a user standpoint, but would be more complicated to convert to a `Duration` cleanly (probably the easiest way is BigDecimal math).

Perhaps all that is outside the intended goal of this PR. Just mentioning the `getChrono` idea since I think that would solve the problem of allowing users and framework developers to manually handle the units the way they want.

File: config/src/main/java/com/typesafe/config/Config.java [hunk 3]
```diff
@@ -794,6 +796,44 @@
...
 
     /**
+     * Gets a value as a java.time.Period. If the value is
+     * already a number, then it's taken as days; if it's
+     * a string, it's parsed understanding units suffixes like
+     * "10d" or "5w" as documented in the <a
+     * href="https://github.com/typesafehub/config/blob/master/HOCON.md">the
+     * spec</a>. This method never returns null.
+     *
+     * @since 1.3.0
```
Comment (marconilanna) [inline config/src/main/java/com/typesafe/config/Config.java]: This should be `@since 1.3.2`

File: config/src/main/java/com/typesafe/config/Config.java [hunk 4]
```diff
@@ -794,6 +796,44 @@
...
+     *             if value cannot be parsed as a number of the given TimeUnit
+     */
+    Period getPeriod(String path);
+
+    /**
+     * Gets a value as a java.time.temporal.TemporalAmount.
+     * This method will first try get get the value as a java.time.Duration, and if unsuccessful,
+     * then as a java.time.Period.
+     * This means that values like "5m" will be parsed as 5 minutes rather than 5 months
+     * @param path path expression
```
Comment (marconilanna) [inline config/src/main/java/com/typesafe/config/Config.java]: Missing the `@since 1.3.2` annotation
